src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -0,0 +1,526 @@
+    ) -> MavenRequirements:
+        field_requirements = (
+            str(maven_coord) for field in fields for maven_coord in (field.value or ())
+        )
+        return MavenRequirements((*field_requirements, *additional_requirements))
+
+
+@dataclass(frozen=True)
+class MavenCoord:
+    """A single Maven-style coordinate for a JVM dependency."""
+
+    # TODO: parse and validate the input into individual coordinate
+    # components, then re-expose the string coordinate as a property
+    # or __str__.
+    coord: str
+
+
+class MavenCoordinates(Collection[MavenCoord]):
+    """An ordered list of MavenCoord."""
+
+
+@dataclass(frozen=True)
+class CoursierLockfileEntry:
+    """A single artifact entry from a Coursier-resolved lockfile.
+
+    These fields are nearly identical to the JSON objects from the
+    "dependencies" entries in Coursier's --json-output-file format.
+    But unlike Coursier's JSON report, a CoursierLockfileEntry
+    includes the content-address of the artifact fetched by Coursier
+    and ingested by Pants.
+
+    For example, a Coursier JSON report dependency entry might look like this:
+
+    ```
+    {
+      "coord": "com.chuusai:shapeless_2.13:2.3.3",
+      "file": "/home/USER/.cache/coursier/v1/https/repo1.maven.org/maven2/com/chuusai/shapeless_2.13/2.3.3/shapeless_2.13-2.3.3.jar",
+      "directDependencies": [
+        "org.scala-lang:scala-library:2.13.0"
+      ],
+      "dependencies": [
+        "org.scala-lang:scala-library:2.13.0"
+      ]
+    }
+    ```
+
+    The equivalent CoursierLockfileEntry would look like this:
+
+    ```
+    CoursierLockfileEntry(
+        coord="com.chuusai:shapeless_2.13:2.3.3", # identical
+        file_name="shapeless_2.13-2.3.3.jar" # PurePath(entry["file"].name)
+        direct_dependencies=(MavenCoord("org.scala-lang:scala-library:2.13.0"),),
+        dependencies=(MavenCoord("org.scala-lang:scala-library:2.13.0"),),
+        file_digest=FileDigest(fingerprint=<sha256 of the jar>, ...),
+    )
+    ```
+    """
+
+    coord: MavenCoord
+    file_name: str
+    direct_dependencies: MavenCoordinates
+    dependencies: MavenCoordinates
+    file_digest: FileDigest
+
+    @classmethod
+    def from_json_dict(cls, entry) -> CoursierLockfileEntry:
+        """Construct a CoursierLockfileEntry from its JSON dictionary representation."""
+
+        return cls(
+            coord=MavenCoord(coord=entry["coord"]),
+            file_name=entry["file_name"],
+            direct_dependencies=MavenCoordinates(
+                MavenCoord(coord=d) for d in entry["directDependencies"]
+            ),
+            dependencies=MavenCoordinates(MavenCoord(coord=d) for d in entry["dependencies"]),
+            file_digest=FileDigest(
+                fingerprint=entry["file_digest"]["fingerprint"],
+                serialized_bytes_length=entry["file_digest"]["serialized_bytes_length"],
+            ),
+        )
+
+    def to_json_dict(self):
```

Comment:
```suggestion
    def to_json_dict(self) -> dict[str, Any]:
```

(requires importing `typing.Any`)